lib/rtti.h

```diff
@@ -0,0 +1,313 @@
+// but mark upper 8 bits to be 0xFF
+template <typename T, typename = void>
+struct TypeIdResolver {
+    static constexpr TypeId resolve() noexcept {
+        // Get original 64-bit hash
+        uint64_t hash = FNV1a(typeNameProxy<T>());
+        // Xor-fold to obtain 56-bit value
+        hash = (hash >> kDiscriminatorBits) ^ (hash & kInnerTypeIdMask);
+        return (kHashDiscriminator << kDiscriminatorBits) | hash;
+    }
+};
+
+// Specialized implementation that looks for static_typeId member.
+template <typename T>
+struct TypeIdResolver<T, std::void_t<decltype(T::static_typeId)>> {
+    static constexpr uint64_t resolve() noexcept {
+        constexpr TypeId id = T::static_typeId();
+        // Allow fall-back implementation if class does not want to explicitly
+        // specify a typeid
+        if constexpr (id == InvalidTypeId) return TypeIdResolver<T, int>::resolve();
+
+        return id;
+    }
+};
+
+}  // namespace detail
+
+/// Given a "full" typeid, returns one with discriminator removed
+static constexpr TypeId innerTypeId(TypeId id) { return id & kInnerTypeIdMask; }
+
+/// Extracts a discriminator from typeid
+static constexpr TypeId typeidDiscriminator(TypeId id) { return id >> kDiscriminatorBits; }
+
+/// Combines typeid and discriminator
+static constexpr TypeId combineTypeIdWithDiscriminator(TypeId discriminator, TypeId id) {
+    return (discriminator << kDiscriminatorBits) | innerTypeId(id);
+}
+
+struct Base;
+
+template <typename This, typename... Parents>
+struct TypeInfo {
+    using T = std::remove_const_t<std::remove_reference_t<This>>;
+
+    static_assert((... && std::is_base_of_v<Parents, This>),
+                  "One or more parents are not a base of this type.");
+
+    static_assert((... && std::is_base_of_v<Base, Parents>),
+                  "One or more parent hierarchies is not based on top of RTTI::Base.");
+
+    static_assert(std::is_base_of_v<Base, This>,
+                  "Invalid typeinfo requested, class is not based on top of RTTI::Base.");
+
+    static_assert(
+        std::is_same_v<T *, decltype(T::rttiEnabledMarker(nullptr))>,
+        "Invalid typeinfo requested, class does not declare typeinfo via DECLARE_TYPEINFO.");
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
```

Comment:
Won't this effectively call the parent `dyn_cast` for all the parents recursively before looking to see if any are non-null?  As soon as a single parent returns a non-null value, you can return that -- no need to check other parents.

Reply:
I tried this and it seems to yield much worse performance. What happens here is that compiler is happily inlines everything (through parents) and we end with quite nice and tight implementation. Adding lazy resolution / early returns here prevents this – too much conditions are added.